cast/src/lib.rs

```diff
@@ -628,6 +629,103 @@ where
         Ok(receipt)
     }
 
+    pub async fn events(&self, hash: String, chain: Chain, api_key: String) -> eyre::Result<String> {
+        let receipt = self.provider.get_transaction_receipt(
+            H256::from_str(&hash)?
+        ).await?.unwrap();
+
+        // Address -> Vector of Events in the contract at that address
+        let mut cached_abis: HashMap<H160, Option<Vec<Event>>> = HashMap::new();
+        let mut return_string: String = String::new();
+
+        for log in receipt.logs {
+            // First, try to grab the ABI from the cache.
+            let abi: Option<Vec<Event>> = if let Some(cached_abi) = cached_abis.get(&log.address) {
+                cached_abi.clone()
+            
+            // If it's not there, go to Etherscan to look up the contract, and parse the interface into an Event ABI.
+            } else {
+                let interfaces = SimpleCast::generate_interface(InterfacePath::Etherscan {
+                    chain: chain,
+                    api_key: api_key.clone(),
+                    address: format!("{:#x}", log.address)
+                        .parse::<Address>()
+                        .wrap_err("Invalid address provided. Did you make a typo?")?,
+                }).await;
+
+                if let Ok(interfaces) = interfaces {
+                    let interface: Vec<Event> = interfaces
+                        .iter()
+                        .map(|iface| iface.source.to_string())
+                        .collect::<Vec<_>>()
+                        .join("\n")
+                        .split("\n")
+                        .map(|line| line.trim().to_string())
+                        .filter(|line| line.starts_with("event"))
```

Comment:
This shouldn't be necessary, since Etherscan client should have access to the abi. Take a look on how `generate_interface` is implemented or `EtherscanIdentifier`.

Reply:
Got it. Question for you about cleanest way to implement it: Looks like it's about 40 lines of code in `generate_interface` to get the ABIs. Doesn't seem ideal to copy all that over. 

Would you abstract that part out into a separate function, then use that function in `generate_interface` and also in `events`? Or is there a better way to approach it?

Reply:
Thanks @joshieDo. Implemented this in latest commit, if you're able to check it out: https://github.com/foundry-rs/foundry/pull/2542/commits/a3371e87ab3197431393613b6ed47bd03f661242